public/parser.js

```diff
@@ -0,0 +1,208 @@
+                    else return { "course": rand, "type": "special" }
+                }
+            }
+        ]));
+
+        $.RULE("courseExpression", () => {
+            let course;
+            let grade = -1;
+
+            course = $.CONSUME(Course);
+            $.OPTION(() => {
+                grade = $.CONSUME(Grade)
+            })
+            if (check) {
+                var courseNum = course.image.match(course.tokenType.PATTERN);
+                // check if this course is good or not
+                // depending on:
+                // 1. taken or not
+                // 2. grade meets minimum
+                if (courseNum != null) {
+                    // do all checks here
+                    for (var taken_course of taken_courses) {
+                        if (courseNum[1].includes(taken_course)) {
+                            console.log(courseNum[1] + " satisfied");
+                            return true;
+                        }
+                    }
+                    return false;
+                }
+                return false;
+            } else {
+                if (grade == -1)
+                    return { "course": course.image, "grade": "" };
+                return { "course": course.image, "grade": grade.image };
+            }
+        })
+
+        $.RULE("parenthesisExpression", () => {
+            let expValue;
+            let grade = "";
+
+            $.CONSUME(LParen);
+            expValue = $.SUBRULE($.andExpression);
+            $.CONSUME(RParen);
+
+            $.OPTION(() => {
+                grade = $.CONSUME(Grade).image;
+            })
+            var res = { "courses": expValue, "grade": grade }
+            if(check) return expValue;
+            else return res
+        });
+
+        // very important to call this after all the rules have been defined.
+        // otherwise the parser may not work correctly as it will lack information
+        // derived during the self analysis phase.
+        this.performSelfAnalysis();
+    }
+}
+
+const parser = new Calculator()
+
+function parseInput(text) {
+    const lexingResult = CalculatorLexer.tokenize(text)
+    // "input" is a setter which will reset the parser's state.
+    parser.input = lexingResult.tokens
+    var res = parser.expression()
+
+    if (parser.errors.length > 0) {
+        throw new Error("sad sad panda, Parsing errors detected")
+    }
+    return res;
+}
+
+function prettyPrint(text) {
+    console.log(text);
+    var res = parseInput(text);
+    console.log(JSON.stringify(res, null, 2));
+    return res;
+}
+
+function verify(text, courses) {
+    check = true;
+    taken_courses = courses;
+    var res = parseInput(text);
+    check = false;
+    taken_courses = [];
+    return res;
+}
+
+    // const inputText = "((CE 2336 or CS 2336 or CS 2337) with a grade of C or better or CS 3333) and (CE 2305 or CS 2305) with a grade of C or better and (CS 2340 or SE 2340 or CE 4304 or EE 4304)"
```

Comment:
Remove the commented code and/or include it in some other form of documentation if it's necessary.
```suggestion
```